## Neural Network with Keras

#### Importing Data

In [1]:
import pandas as pd

In [12]:
heart_df = pd.read_csv('heart_disease.csv')
heart_df.head()

Unnamed: 0   age  sex  chest_pain  resting_blood_pressure  \
0           0  70.0  1.0         4.0                   130.0   
1           1  67.0  0.0         3.0                   115.0   
2           2  57.0  1.0         2.0                   124.0   
3           3  64.0  1.0         4.0                   128.0   
4           4  74.0  0.0         2.0                   120.0   

   serum_cholestoral  fasting_blood_sugar  resting_ecg_results  \
0              322.0                  0.0                  2.0   
1              564.0                  0.0                  2.0   
2              261.0                  0.0                  0.0   
3              263.0                  0.0                  0.0   
4              269.0                  0.0                  2.0   

   max_heart_rate_achieved  exercise_induced_angina  oldpeak  \
0                    109.0                      0.0      2.4   
1                    160.0                      0.0      1.6   
2                    141.0                      0.0      0.3   
3                    105.0                      1.0      0.2   
4                    121.0                      1.0      0.2   

   slope of the peak  num_of_major_vessels  thal  heart_disease  
0                2.0                   3.0   3.0              2  
1                2.0                   0.0   7.0              1  
2                1.0                   0.0   7.0              2  
3                2.0                   1.0   7.0              1  
4                1.0                   1.0   3.0              1

In [13]:
heart_df.columns

Index(['Unnamed: 0', 'age', 'sex', 'chest_pain', 'resting_blood_pressure',
       'serum_cholestoral', 'fasting_blood_sugar', 'resting_ecg_results',
       'max_heart_rate_achieved', 'exercise_induced_angina', 'oldpeak',
       'slope of the peak', 'num_of_major_vessels', 'thal', 'heart_disease'],
      dtype='object')

In [14]:
heart_df = heart_df.drop(columns = ['Unnamed: 0'])
heart_df.head()

age  sex  chest_pain  resting_blood_pressure  serum_cholestoral  \
0  70.0  1.0         4.0                   130.0              322.0   
1  67.0  0.0         3.0                   115.0              564.0   
2  57.0  1.0         2.0                   124.0              261.0   
3  64.0  1.0         4.0                   128.0              263.0   
4  74.0  0.0         2.0                   120.0              269.0   

   fasting_blood_sugar  resting_ecg_results  max_heart_rate_achieved  \
0                  0.0                  2.0                    109.0   
1                  0.0                  2.0                    160.0   
2                  0.0                  0.0                    141.0   
3                  0.0                  0.0                    105.0   
4                  0.0                  2.0                    121.0   

   exercise_induced_angina  oldpeak  slope of the peak  num_of_major_vessels  \
0                      0.0      2.4                2.0                   3.0   
1                      0.0      1.6                2.0                   0.0   
2                      0.0      0.3                1.0                   0.0   
3                      1.0      0.2                2.0                   1.0   
4                      1.0      0.2                1.0                   1.0   

   thal  heart_disease  
0   3.0              2  
1   7.0              1  
2   7.0              2  
3   7.0              1  
4   3.0              1

In [15]:
heart_df.shape

(270, 14)

There are 270 observations. This means that your neural network will have an input data of shape 270 x 13, excluding the target variable (heart_disease). The features present in the dataset are:

- age
- sex
- chest pain type (4 values)
- resting blood pressure
- serum cholesterol in mg/dl
- fasting blood sugar > 120 mg/dl
- resting electrocardiographic results (values 0,1,2)
- maximum heart rate achieved
- exercise-induced angina
- oldpeak (ST depression induced by exercise relative to rest)
- the slope of the peak exercise ST segment
- number of major vessels (0–3) colored by fluoroscopy
- thal (3 = normal; 6 = fixed defect; 7 = reversible defect)
- heart_disease: absence (1) or presence (2) of heart disease

Next, you can check for missing values and also the data types. A Neural Network expects all features to be numeric and not contain missing values.

In [16]:
heart_df.isna().sum()

age                        0
sex                        0
chest_pain                 0
resting_blood_pressure     0
serum_cholestoral          0
fasting_blood_sugar        0
resting_ecg_results        0
max_heart_rate_achieved    0
exercise_induced_angina    0
oldpeak                    0
slope of the peak          0
num_of_major_vessels       0
thal                       0
heart_disease              0
dtype: int64

In [17]:
heart_df.dtypes

age                        float64
sex                        float64
chest_pain                 float64
resting_blood_pressure     float64
serum_cholestoral          float64
fasting_blood_sugar        float64
resting_ecg_results        float64
max_heart_rate_achieved    float64
exercise_induced_angina    float64
oldpeak                    float64
slope of the peak          float64
num_of_major_vessels       float64
thal                       float64
heart_disease                int64
dtype: object

The target variableheart_disease has values 1 and 2: absence (1) or presence (2) of heart disease. We change it to 0 and 1. Value 1 means "have heart disease" and 0 means "do not have heart disease".

In [18]:
#replace target class with 0 and 1 
#1 means "have heart disease" and 0 means "do not have heart disease"
heart_df['heart_disease'] = heart_df['heart_disease'].replace(1, 0)
heart_df['heart_disease'] = heart_df['heart_disease'].replace(2, 1)

In [19]:
heart_df.head()

age  sex  chest_pain  resting_blood_pressure  serum_cholestoral  \
0  70.0  1.0         4.0                   130.0              322.0   
1  67.0  0.0         3.0                   115.0              564.0   
2  57.0  1.0         2.0                   124.0              261.0   
3  64.0  1.0         4.0                   128.0              263.0   
4  74.0  0.0         2.0                   120.0              269.0   

   fasting_blood_sugar  resting_ecg_results  max_heart_rate_achieved  \
0                  0.0                  2.0                    109.0   
1                  0.0                  2.0                    160.0   
2                  0.0                  0.0                    141.0   
3                  0.0                  0.0                    105.0   
4                  0.0                  2.0                    121.0   

   exercise_induced_angina  oldpeak  slope of the peak  num_of_major_vessels  \
0                      0.0      2.4                2.0                   3.0   
1                      0.0      1.6                2.0                   0.0   
2                      0.0      0.3                1.0                   0.0   
3                      1.0      0.2                2.0                   1.0   
4                      1.0      0.2                1.0                   1.0   

   thal  heart_disease  
0   3.0              1  
1   7.0              0  
2   7.0              1  
3   7.0              0  
4   3.0              0

Now the data is ready. There are no missing values in the dataset, and all features are numeric. Next, you’ll separate the target from the data, split into train and test set, and then standardize the data.

In [23]:
# split into input (X) and output (y) variables
X = heart_df.iloc[:,0:13]
y = heart_df.iloc[:,13]

In [24]:
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

In [28]:
#split data into train and test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=2)

#standardize the dataset
sc = StandardScaler()
sc.fit(X_train)
X_train = sc.transform(X_train)
X_test = sc.transform(X_test)

In [29]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((202, 13), (68, 13), (202,), (68,))

#### NN with Keras

In [43]:
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam

model = Sequential()
model.add(Dense(12, input_dim=13, activation='relu'))
model.add(Dense(8, activation='relu'))
model.add(Dense(1, activation='sigmoid'))

In [56]:
# compile the model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
model.fit(X_train, y_train, epochs=100, verbose=0)

In [57]:
train_acc = model.evaluate(X_train, y_train, verbose=0)[1]
test_acc = model.evaluate(X_test, y_test, verbose=0)[1]

print("Train accuracy of keras neural network: {}".format(round((train_acc * 100), 2)))
print("Test accuracy of keras neural network: {}".format(round((test_acc * 100),2)))

Train accuracy of keras neural network: 100.0
Test accuracy of keras neural network: 82.35


In [70]:
# make class predictions with the model
predictions = model.predict_classes(X_test)
predictions

array([[1],
       [0],
       [1],
       [1],
       [0],
       [0],
       [1],
       [0],
       [1],
       [1],
       [0],
       [1],
       [1],
       [0],
       [1],
       [0],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [0],
       [0],
       [0],
       [1],
       [1],
       [0],
       [1],
       [1],
       [1],
       [1],
       [0],
       [1],
       [0],
       [1],
       [0],
       [1],
       [1],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [1],
       [1],
       [0],
       [0],
       [0],
       [0],
       [0],
       [1],
       [0],
       [0],
       [0],
       [1],
       [0],
       [0],
       [0],
       [1],
       [1],
       [0],
       [0],
       [0]], dtype=int32)

### Another Data 'pima_diabetes.csv'

#### Loading Data

In [72]:
import pandas as pd
dataset = pd.read_csv('pima_diabetes.csv')
dataset.head()

6  148  72  35    0  33.6  0.627  50  1
0  1   85  66  29    0  26.6  0.351  31  0
1  8  183  64   0    0  23.3  0.672  32  1
2  1   89  66  23   94  28.1  0.167  21  0
3  0  137  40  35  168  43.1  2.288  33  1
4  5  116  74   0    0  25.6  0.201  30  0

#### About dataset
Attributes: 
 1. Number of times pregnant
 2. Plasma glucose concentration a 2 hours in an oral glucose tolerance test
 3. Diastolic blood pressure (mm Hg)
 4. Triceps skin fold thickness (mm)
 5. 2-Hour serum insulin (mu U/ml)
 6. Body mass index (weight in kg/(height in m)^2)
 7. Diabetes pedigree function
 8. Age (years)
 9. Class variable (0 or 1). Class Distribution: (class value 1 is interpreted as "tested positive for
   diabetes")

Let us name the columns using numbers 1:9 as above.

In [73]:
# add header names
headers =  ['1', '2', '3', '4', '5', '6', '7', '8', '9']

dataset = pd.read_csv('pima_diabetes.csv', names= headers)
dataset.head()

1    2   3   4    5     6      7   8  9
0  6  148  72  35    0  33.6  0.627  50  1
1  1   85  66  29    0  26.6  0.351  31  0
2  8  183  64   0    0  23.3  0.672  32  1
3  1   89  66  23   94  28.1  0.167  21  0
4  0  137  40  35  168  43.1  2.288  33  1

In [74]:
dataset.shape

(768, 9)

In [75]:
dataset.isna().sum()

1    0
2    0
3    0
4    0
5    0
6    0
7    0
8    0
9    0
dtype: int64

In [76]:
dataset.dtypes

1      int64
2      int64
3      int64
4      int64
5      int64
6    float64
7    float64
8      int64
9      int64
dtype: object

##### Looks everything good. Next, we separate the target from the data, split into train and test set, and then standardize the data.

In [81]:
# split into input (X) and output (y) variables
X = dataset.iloc[:,0:8]
y = dataset.iloc[:,8]

In [82]:
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

In [83]:
#split data into train and test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=2)

#standardize the dataset
sc = StandardScaler()
sc.fit(X_train)
X_train = sc.transform(X_train)
X_test = sc.transform(X_test)

In [84]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((576, 8), (192, 8), (576,), (192,))

In [152]:
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam

#### What Is a Sample?

A sample is a single row of data. A training dataset is comprised of many rows of data, e.g. many samples. A sample may also be called an instance, an observation, an input vector, or a feature vector.

#### What Is a Batch?

The batch size is a hyperparameter that defines the number of samples to work through before updating the internal model parameters. Like a for-loop iterating over one or more samples and making predictions.

A training dataset can be divided into one or more batches.
 - Batch Gradient Descent. Batch Size = Size of Training Set
 - Stochastic Gradient Descent. Batch Size = 1
 - Mini-Batch Gradient Descent. 1 < Batch Size < Size of Training Set
 
In the case of mini-batch gradient descent, popular batch sizes include 32, 64, and 128 samples.

#### What Is an Epoch?

The number of epochs is a hyperparameter that defines the number of times that the learning algorithm will work through the entire training dataset.

One epoch means that each sample in the training dataset has had an opportunity to update the internal model parameters. An epoch is comprised of one or more batches.

The number of epochs is traditionally large, often hundreds or thousands, allowing the learning algorithm to run until the error from the model has been sufficiently minimized. You may see examples of the number of epochs in the literature and in tutorials set to 10, 100, 500, 1000, and larger.

#### What Is the Difference Between Batch and Epoch?

The batch size is a number of samples processed before the model is updated.

The number of epochs is the number of complete passes through the training dataset.

The size of a batch must be more than or equal to one and less than or equal to the number of samples in the training dataset.


In [230]:

model = Sequential()
# The number of hidden neurons should be less than twice the size of the input layer neurons.
# The number of input varible is the input_dim, which is the number of the input layer neurons.
# choose number of hidden layers and number of neurons in hidden layers by trial and error method
# first hidden layer having 12 nodes with activation function'relu'
model.add(Dense(12, input_dim = 8, activation='relu'))  
# second hidden layer having 6 nodes with activation function 'relu'
model.add(Dense(8, activation='relu')) 
# Output layer having 1 node with activation function 'sigmoid'
model.add(Dense(1, activation='sigmoid'))

In [286]:
# compile the model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
# choose epochs and batch_size as trial and error method
model.fit(X_train, y_train, epochs= 1000, verbose=0, batch_size=32)

In [287]:
train_acc = model.evaluate(X_train, y_train, verbose=0)[1]
test_acc = model.evaluate(X_test, y_test, verbose=0)[1]

print("Train accuracy of keras neural network: {}".format(round((train_acc * 100), 2)))
print("Test accuracy of keras neural network: {}".format(round((test_acc * 100),2)))

Train accuracy of keras neural network: 99.13
Test accuracy of keras neural network: 74.48


In [288]:
# evaluate the keras model
_, accuracy = model.evaluate(X_test, y_test)
print('Accuracy: %.2f' % (accuracy*100))

6/6 [==============================] - 0s 1ms/step - loss: 7.1907 - accuracy: 0.7448
Accuracy: 74.48


In [289]:
# make class predictions with the model
predictions = model.predict_classes(X_test)
predictions

array([[0],
       [0],
       [0],
       [1],
       [1],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [1],
       [0],
       [0],
       [0],
       [1],
       [1],
       [1],
       [0],
       [1],
       [0],
       [0],
       [1],
       [1],
       [0],
       [0],
       [1],
       [1],
       [0],
       [0],
       [0],
       [1],
       [1],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [1],
       [0],
       [0],
       [0],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [0],
       [1],
       [0],
       [1],
       [1],
       [1],
       [1],
       [0],
       [1],
       [1],
       [0],
       [0],
       [0],
       [0],
       [1],
       [0],
       [0],
       [0],
       [1],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
    